src/Uno.UWP/Globalization/ApplicationLanguages.cs

```diff
@@ -20,16 +21,71 @@ public static partial class ApplicationLanguages
 		static ApplicationLanguages()
 		{
 #if !NET461
-			if (ApplicationData.Current.LocalSettings.Values.TryGetValue(PrimaryLanguageOverrideSettingKey, out var savedValue)
-				&& savedValue is string stringSavedValue)
+			var appSpecificKey = GetAppSpecificSettingKey();
+			var hasLegacySetting = TryGetSettingFromKey(PrimaryLanguageOverrideSettingKey, out var legacySavedValue);
+			var hasAppSpecificSetting = TryGetSettingFromKey(appSpecificKey, out var appSpecificSavedValue);
+
+			// We have four cases:
+			// 1. hasAppSpecificSetting && hasLegacySetting
+			// 2. hasAppSpecificSetting && !hasLegacySetting
+			// 3. !hasAppSpecificSetting && hasLegacySetting
+			// 4. !hasAppSpecificSetting && !hasLegacySetting
+
+			// For 1 and 2, we will ignore the legacy setting and use the app specific setting.
+			// For 3, we will copy the setting from legacy to app specific
+			// For 4, we will do nothing.
+
+			// History:
+			// When persistent was implemented, it relied on saving the PrimaryLanguageOverride value to
+			// "__Uno.PrimaryLanguageOverride" in LocalSettings of ApplicationData.
+			// After that, we found that in Skia, the LocalSettings of ApplicationData is shared across all applications (which is wrong).
+			// The fix for having app-specific local settings on Skia is being done for Uno 5 (see https://github.com/unoplatform/uno/pull/12314)
+			// However, since the above is breaking change, we won't take it for Uno 4.x, but we still want PrimaryLanguageOverride to behave properly in Uno 4.x.
+			// So, we modify the key to have an app-specific part.
+			// We try to copy the value we read from the "legacy" key to the app-specific key.
+			// In a future major release, we may consider removing the legacy setting key.
+
+			// Note: There is a Wasm concern that is mentioned in https://github.com/unoplatform/uno/issues/12320
+			// So, we decided to implement this for all platforms.
+
+			if (hasAppSpecificSetting)
 			{
-				_primaryLanguageOverride = stringSavedValue;
+				_primaryLanguageOverride = appSpecificSavedValue;
+			}
+			else if (hasLegacySetting)
+			{
+				ApplicationData.Current.LocalSettings.Values[appSpecificKey] = legacySavedValue;
+				_primaryLanguageOverride = legacySavedValue;
 			}
 #endif
 
 			ApplyLanguages();
 		}
 
+#if !NET461
+		private static bool TryGetSettingFromKey(string key, out string value)
+		{
+			if (ApplicationData.Current.LocalSettings.Values.TryGetValue(key, out var savedValue))
+			{
+				value = savedValue as string;
+				return value is not null;
+			}
+
+			value = null;
+			return false;
+		}
+
+		private static string GetAppSpecificSettingKey()
+		{
+			if (Package.EntryAssembly is not { } entryAssembly)
+			{
+				throw new InvalidOperationException("ApplicationLanguages is being accessed too early before an instance of Application was created.");
+			}
+
+			return $"PrimaryLanguageOverrideSettingKey.{entryAssembly.GetName().Name}";
```

Comment:
I don't believe this should be the fix. The origin of the issue is likely because we're not storing app settings in a app identity based location, but rather a common location that is incorrectly shared.

Reply:
Right but how do we fix this in 4.x without it being a breaking change (ie what @MartinZikmund is working on).
My additional comment would be that this fix would limit overlap of wasm apps when debugging - currently if you use the same port (the template defaults to a fixed post to allow ui tests to "just work") you'll see PrimaryLanguageOverride from other apps you debugged with the same port. Adding the app name would limit this overlap

Reply:
Those are two different issues then. If this fix is for wasm, we're talking about a debug issue (which I don't believe needs fixing), not a production issue. The issue for Skia is different, where settings are placed at the wrong location.

In general linking the settings to the assembly name will cause issues, as if an app is logically deployed at the same location, but somehow changes its internal structure, the settings will be lost.

Reply:
That's indeed interesting. Per [the documentation](https://www.w3schools.com/html/html5_webstorage.asp): _Web storage is per origin (per domain and protocol). All pages, from one origin, can store and access the same data._

I don't think that we should be discriminating based on the assembly name, though. We're currently not using the app manifest for wasm, but it feels like we should start doing so, so we can pick up the app identity and use it as a based for storing settings.

This is likely something we should be explaining in the storage docs as well, as files from one app may interfere in another app from the same origin.

/cc @MartinZikmund